diff --git a/aya-obj/src/btf/btf.rs b/aya-obj/src/btf/btf.rs
index de9e54e..c7ae2be 100644
--- a/aya-obj/src/btf/btf.rs
+++ b/aya-obj/src/btf/btf.rs
@@ -676,6 +676,8 @@ impl Btf {
                             }
                             ty.set_linkage(FuncLinkage::Static);
                         }
+                    } else {
+                        ty.set_linkage(FuncLinkage::Static);
                     }
                 }
                 // Sanitize FLOAT.
